helix-core/src/test/java/org/apache/helix/integration/rebalancer/WagedRebalancer/TestWagedLoadedCluster.java

```diff
@@ -0,0 +1,277 @@
+      ImmutableMap.of("CU", 50, "DISK", 50);
+
+  private final Map<String, Integer> _defaultPartitionWeight =
+      ImmutableMap.of("CU", 10, "DISK", 10);
+
+  private final Map<String, Integer> _newPartitionWeight =
+      ImmutableMap.of("CU", 20, "DISK", 10);
+  private final int DEFAULT_DELAY = 500; // 0.5 second
+  private static final Logger LOG = LoggerFactory.getLogger(CLASS_NAME);
+
+  @BeforeClass
+  public void beforeClass() throws Exception {
+    LOG.info("START " + CLASS_NAME + " at " + new Date(System.currentTimeMillis()));
+    _gSetupTool.addCluster(CLUSTER_NAME, true);
+    // create 6 node cluster
+    for (int i = 0; i < NUM_NODE; i++) {
+      String storageNodeName = PARTICIPANT_PREFIX + "_" + (START_PORT + i);
+      _gSetupTool.addInstanceToCluster(CLUSTER_NAME, storageNodeName);
+      _nodes.add(storageNodeName);
+    }
+    // ST downward message will get delayed by 5sec.
+    startParticipants(DEFAULT_DELAY);
+
+    // start controller
+    String controllerName = CONTROLLER_PREFIX + "_0";
+    _controller = new ClusterControllerManager(ZK_ADDR, CLUSTER_NAME, controllerName);
+    _controller.syncStart();
+
+    enablePersistBestPossibleAssignment(_gZkClient, CLUSTER_NAME, true);
+
+    _assignmentMetadataStore =
+        new AssignmentMetadataStore(new ZkBucketDataAccessor(ZK_ADDR), CLUSTER_NAME) {
+          public Map<String, ResourceAssignment> getBaseline() {
+            // Ensure this metadata store always read from the ZK without using cache.
+            super.reset();
+            return super.getBaseline();
+          }
+
+          public synchronized Map<String, ResourceAssignment> getBestPossibleAssignment() {
+            // Ensure this metadata store always read from the ZK without using cache.
+            super.reset();
+            return super.getBestPossibleAssignment();
+          }
+        };
+
+    // Set test instance capacity and partition weights
+    HelixDataAccessor dataAccessor = new ZKHelixDataAccessor(CLUSTER_NAME, _baseAccessor);
+    ClusterConfig clusterConfig =
+        dataAccessor.getProperty(dataAccessor.keyBuilder().clusterConfig());
+
+    clusterConfig.setDefaultInstanceCapacityMap(_defaultInstanceCapacity);
+    clusterConfig.setDefaultPartitionWeightMap(_defaultPartitionWeight);
+    dataAccessor.setProperty(dataAccessor.keyBuilder().clusterConfig(), clusterConfig);
+
+    // Create 3 resources with 2 partitions each.
+    for (int i = 0; i < 3; i++) {
+      String db = "Test-WagedDB-" + i;
+      createResourceWithWagedRebalance(CLUSTER_NAME, db, BuiltInStateModelDefinitions.MasterSlave.name(),
+          2 /*numPartitions*/, 3 /*replicas*/, 3 /*minActiveReplicas*/);
+      _gSetupTool.rebalanceStorageCluster(CLUSTER_NAME, db, 3);
+      _allDBs.add(db);
+    }
+
+    // Start a thread which will keep validating instance usage using currentState and pending messages.
+    Thread validateInstanceUsageThread = new Thread(new Runnable() {
```

Comment:
Would suggest create as a variable and stop it in after class. At least, set it as daemon. Otherwise, could be thread leakage.

Reply:
done.